File: 백준/Platinum/16930. 달리기/달리기.py
import sys
from collections import deque
input = sys.stdin.readline

# 1초에 이동 가능한 최대 이동거리 K
row, col, K = map(int, input().split())
graph = [list(map(str,input().strip())) for _ in range(row)]
visited = [[0]*col for _ in range(row)]

# depart: x1, y1 / destination: x2, y2
x1, y1, x2, y2 = map(int, input().split())

# 인덱스 접근 위해 -1
queue = deque()
queue.append((x1-1,y1-1))

while queue:
    x, y = queue.popleft()
    
    if x == x2-1 and y == y2-1:
        print(visited[x][y])
        exit()
        
    # 델타탐색
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]

    for i in range(4):
        for multiplier in range(1, K+1):
            nx = x + (multiplier*dx[i])
            ny = y + (multiplier*dy[i])
            
            if nx < 0 or nx >= row or ny < 0 or ny >= col:
                break
            if graph[nx][ny] == '#':
                break
            if visited[nx][ny] == 0:
                queue.append((nx,ny))
                visited[nx][ny] = visited[x][y] + 1
            elif visited[nx][ny] == visited[x][y] + 1:
                continue
            else:
                break
print(-1)

File: 백준/Platinum/16930. 달리기/달리기.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static int M;
    static int K;
    static int endX;
    static int endY;
    static String[][] graph;
    static int[][] visited;

    static class Coord {
        int x;
        int y;
        Coord(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken()); // 1초에 이동가능 최대 거리

        visited = new int[N+1][M+1]; // 방문여부 및 소요시간 저장
        graph = new String[N+1][M+1]; // 체육관 구조
        for(int i = 1; i < N + 1; i++) {
            String[] input = br.readLine().split("");
            for(int j = 1; j < M + 1; j++) {
                graph[i][j] = input[j-1];
            }
        }

        st = new StringTokenizer(br.readLine());
        br.close();

        int startX = Integer.parseInt(st.nextToken());
        int startY = Integer.parseInt(st.nextToken());
        endX = Integer.parseInt(st.nextToken());
        endY = Integer.parseInt(st.nextToken());

        System.out.println(bfs(new Coord(startX, startY)));
    }

    static int bfs(Coord coordinate) {

        if(coordinate.x == endX && coordinate.y == endY) {
            return 0;
        }

        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, 1, -1};
        Queue<Coord> queue = new LinkedList<>();
        queue.add(coordinate);

        while(!queue.isEmpty()) {
            Coord currentCoord = queue.poll();
            int currentX = currentCoord.x;
            int currentY = currentCoord.y;

            for(int i = 0; i < 4; i++) { // 델타탐색
                for(int j = 1; j < K + 1; j++) { // i방향으로 1부터 k값만큼 거리이동
                    int nx = currentX + (j * dx[i]);
                    int ny = currentY + (j * dy[i]);

                    if(nx < 1 || nx > N || ny < 1 || ny > M) { // 범위 벗어나지 않는지 체크
                        break;
                    }
                    if(graph[nx][ny].equals("#")) { // 벽인지 체크
                        break;
                    }
                    if(visited[nx][ny] == 0) { // 방문하지 않았다면 queue에 추가 및 시간 저장

                        visited[nx][ny] = visited[currentX][currentY] + 1;
                        // 목적지 도달 시에 소요시간 출력
                        if(nx == endX && ny == endY) {
                            return visited[nx][ny];
                        }
                        queue.add(new Coord(nx, ny));
                    }
                    // 재방문의 경우 방문 좌표의 값이 기존 좌표의 값보다 작으면 더 이상 진해할 필요 없음
                    else if(visited[nx][ny] <= visited[currentX][currentY]) {
                        break;
                    }
                }
            }
        }
        return -1;
    }
}